### ISSUE TITLE
Support for `.rb` processors that match on the file path

### ISSUE THREAD

**Issue by joeldrapper** [camertron/onload#2]
Would you be interested in supporting processors for normal `.rb` Ruby files matched on the file path rather than the file extension?

The interface could be something like this:

```ruby
class UpcasePreprocessor
  def self.call(source, path)
    if path.start_with?("foo")
      source.gsub(/(\"\w+\")/, '\1.upcase')
    end
  end
end
```

**Comment by joeldrapper** [comment]
Another option that might help avoid breaking changes in the future would be to pass an `Onload::File` Data object to the preprocessor `#call` method. Preprocessors can use `file.source` for now, but going forward, that file object could give the preprocessors access to more features such as `file.path`.

**Comment by camertron** [comment]
Onload is [hard-coded](https://github.com/camertron/onload/blob/764533fab3be136d21aba37606a3a6699a8861d4/lib/onload/file.rb#L23) to write .rb files, so right now this isn't possible. However I don't think there's any technical reason it couldn't support processing .rb files. Your suggestion of passing the path or even an instance of `Onload::File` to `.call` makes sense, although I wonder if another method on the processor might work?

```ruby
class UpcasePreprocessor
  def self.process?(path)
    path.start_with("foo")
  end
end
```

The bigger issue IMO is where to write the output file, since `Kernel.require` and `Kernel.load` both operate on paths. There might be something to letting users provide their own `Onload::File` subclasses. They could override the `#outfile` method.

**Comment by joeldrapper** [comment]
Oh I see. I’ll be honest, I haven't read the source code but my assumption was that it was generating Ruby dynamically and loading that without writing it anywhere. I wonder if it would be possible to use eval to avoid writing to a file.

**Comment by camertron** [comment]
Oh yeah no problem. It more or less has to write a file in order to delegate to Ruby's original `require`. `eval` could work, but the problem with skipping the normal `require` method is that other gems like Zeitwerk, bootsnap, etc are effectively skipped as well, since they also rely on monkeypatching `require`.

**Comment by joeldrapper** [comment]
Yeah, that’s tricky. I guess you could temporarily overwrite the file with the processed contents, require it and then put the original contents back. That sounds like something that might be quite slow but it might be worth trying.

**Comment by joeldrapper** [comment]
Another approach I’ve been prototyping is using TracePoint’s `script_compiled` event to hook in right after a require, and then use eval to overwrite each method definition with processed method definitions from the parsed file. This approach is quite limited and it's not very efficient, but in some ways it's quite a bit less invasive since it doesn't require any patches to work.

**Comment by camertron** [comment]
Yeah, overwriting the contents of the file and setting it back again could be a good way to go. You'd have to be careful to make sure the original contents get restored if the user mashes Ctrl+C in the middle of execution 😅 Another option could be sticking a temp directory to the front of the load path so it requires everything from there instead of the original location.

**Comment by joeldrapper** [comment]
Using a `temp` directory in the load path would work for `require` but not `require_relative`. It would also mess up stack traces. We could add some extra safety by writing a backup to a temporary file.

1. Load and parse the original Ruby file
2. Write the original contents to a temp file, e.g. `filename.rb.tmp`
3. Write the processed contents to the original file
4. Call `super`
5. Write the original contents back to the original file
6. Delete the temp file

If we keep a global register, we could also rescue `SystemExit` and ensure all the original files are put back.

**Comment by joeldrapper** [comment]
@camertron I explored using `eval` and `RubyVM::InstructionSequence.compile_file(file).eval` from a `require` patch. Neither of these techniques worked when the file defined a constant that had been registered to autoload.

But I was able to get it to work by calling `super` after temporarily replacing the contents of the file. Here’s my prototype:

```ruby
def require(feature)
  type, path = $LOAD_PATH.resolve_feature_path(feature)
  if path&.start_with?(__dir__)
    begin
      original_contents = File.read(path)
      File.write(path, "puts 'Hello';" + original_contents)
      super
    ensure
      File.write(path, original_contents)
    end
  else
    super
  end
end
```

I’ll keep hacking on a way to load source without having to replace the original file contents, even when it creates an autoload constant.
